### ISSUE TITLE
Fatal crash during binder generation

### ISSUE THREAD

**Issue by trelau** [RosettaCommons/binder#23]
Is there any way to get diagnostic output when I get a fatal crash?

![image](https://cloud.githubusercontent.com/assets/8950602/26449329/8b0f54d6-411f-11e7-8778-9843b2b00e34.png)

I didn't build in debug mode, I'll try that next, but just wondering if there is a verbosity flag or anything to see what it's working on before/during crash. It reads the thousands of header files in all_includes.h, but then crashes during the binder generation before writing any output files. Based on some other issues I've read here, my guess it has something to do with errors in my header files (forward declarations and order of include files).

**Comment by lyskov** [comment]
Hmm... any errors in headers should result in diagnostic printout similar to what clang do on compiling the same headers. There is no special options i can think of for this kind of situation (all errors that already detectable by Binder will should be printed already) but there is a few things that you can try before switching to debug build: 
 - check the memory consumption of Binder process, if you put a lot of headers into all-includes.hh then parsing could easily take take considerable amount of memory (for our project (Rosetta) Binder consume ~4Gb during this phase)
 - from the terminal log copy Binder command line and replace 'binder' in it with 'clang++ and see if that works without the errors

**Comment by trelau** [comment]
Looks like it peaks around 500Mb of memory. For the second part, if I understood you correctly, I used clang++ pointed to all_includes.h, leaving include directories but removing binder specific options. In that case, it throws one error about a struct being redefined in two different files. After fixing that it doesn't throw any errors. When I run binder if spits out a bunch of "note: in instantiation of member function ...". Seems to happen at templates.

I now see there are a lot of forward declarations used in the include files which seem to be a problem based on this documentation. Maybe replace all forward declarations with the designated #include file?

If I didn't mention it before I'm on Windows 10 in case there are any extra I should be passing in.

**Comment by lyskov** [comment]
What do you mean by ' there are a lot of forward declarations used in the include files'? Also, could you please point me to where forward declaration is consider to be problem? I am not aware of any of such issues...

So after struct redefinition was fixed - does Binder run continue to terminate abnormally? If so, would you be interested to try to isolate declaration that lead to this? What i have in mind is to try to use binary search and start removing include from all-include.hh until you found what trigger this. Alternatively if you can produce stack trace for debug build it might helpful for me as well.

Is this proprietary code or something that can download and try to replicate locally?

**Comment by trelau** [comment]
Hmm...I'm not a C++ expert so perhaps my terminology is incorrect, but I was referring to this thread:

https://github.com/RosettaCommons/binder/issues/11#issuecomment-282100948

And I thought "forward declared" meant there was code like "class Foo;" instead of #include <Foo.h> in other include files.

Yes Binder continues to terminate. This is something I've been internally experimenting with to learn some new skills, perhaps I should tackle a much smaller project. I'll experiment a little more and report any breakthroughs, but perhaps it's not a Binder issue and I'll close the ticket accordingly.

**Comment by trelau** [comment]
So after some debugging I commented out an exception and uncommented some code allowing a forward declaration. This was in context.cpp:

![image](https://cloud.githubusercontent.com/assets/8950602/26564774/4f9f5a4e-44b3-11e7-8292-66684ff973a1.png)

See here for original code: https://github.com/RosettaCommons/binder/blob/master/binder/context.cpp#L254

After this binder would run and generate the bindings. Although, the bindings won't compile but I don't think it's a binder or pybind11 issue, but rather an issue with the library being wrapped. I'll go ahead and close this issue for now. Thanks for all the support.

**Comment by lyskov** [comment]
Thank you for letting me know @trelau! 

A few questions: in your initial description you simply mentioned that there was a 'fatal crash': have you seen message `Could not find binder for type` in the logs before the crash? (i am trying to figure out if i need to refactor this code to make it more clean for user whats is going on etc)

And second: please note that commented code that generate `forward declaration` will no longer work with recent Pybind11 (it will terminate on import). So i think the only solution here is to make sure that all-include.hh for your project is indeed contain definition of all types.

And lastly: in most cases Binder will just avoid binding functions that arguments thats not fully declared. But this does not always work, especially for template cases.

Hope this helps,

**Comment by trelau** [comment]
Hmm...I don't remember seeing the "Could not find binder for type..." anywhere before or during the fatal crash. It wasn't until I debugged and stepped into the code did I see it terminating at that line and decided to uncomment the other parts. And yes, I experienced the forward declaration issue after forcing a build and trying to import into Python. I only saw the forward declaration warning for 3 out of the thousands of classes and methods I attempted to wrap. Although I ran into missing headers and types in the binders and other pybind11 issues that would need more effort to resolve. The library I was experimenting with makes heavy use of templates.
